src/realm_dart_logger.cpp

```diff
@@ -0,0 +1,113 @@
+////////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2023 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////////
+
+#include <sstream>
+#include <set>
+#include <mutex>
+#include <thread>
+#include <map>
+#include <chrono>
+
+#include "realm_dart_logger.h"
+#include <realm/object-store/c_api/types.hpp>
+
+auto& dart_logger_mutex = *new std::mutex;
+bool is_core_logger_callback_set = false;
+std::map<Dart_Port, realm_log_level_e> dart_send_ports;
+realm_log_level_e default_log_level = RLM_LOG_LEVEL_INFO;
+
+
+bool isPortRegistered(Dart_Port key)
+{
+    return (dart_send_ports.find(key) != dart_send_ports.end());
+}
+
+realm_log_level_e calucale_minimum_log_level()
+{
+    auto min_log_level = default_log_level;
```

Comment:
calucale_minimum_log_level is an internal function that is called by other functions, which are already guarded.

Reply:
This could pose a problem if someone refactors the code later. Either we guard it as well or we should have a comment that it needs to be invoked with the lock held. The later option is not ideal but would work

Reply:
// This function should be invoked while holding the lock `dart_logger_mutex`

Reply:
we can also try std::recursive_mutex